services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java

```diff
@@ -23,14 +23,25 @@
  *
  * TODO: If later on DDS router could support a better way to register a handler to handle the exceptional cases,
  * we should update the logic here.
- *
- * TODO: Remove all {@link Optional} from this class.
  */
 
 public class RouterExceptionAndTrackingUtils {
   public enum FailureType {
-    REGULAR, SMART_RETRY_ABORTED_BY_SLOW_ROUTE, SMART_RETRY_ABORTED_BY_DELAY_CONSTRAINT,
-    SMART_RETRY_ABORTED_BY_MAX_RETRY_ROUTE_LIMIT, RESOURCE_NOT_FOUND, RETRY_ABORTED_BY_NO_AVAILABLE_REPLICA
+    REGULAR(null), SMART_RETRY_ABORTED_BY_SLOW_ROUTE(AggRouterHttpRequestStats::recordSlowRouteAbortedRetryRequest),
+    SMART_RETRY_ABORTED_BY_DELAY_CONSTRAINT(AggRouterHttpRequestStats::recordDelayConstraintAbortedRetryRequest),
+    SMART_RETRY_ABORTED_BY_MAX_RETRY_ROUTE_LIMIT(AggRouterHttpRequestStats::recordRetryRouteLimitAbortedRetryRequest),
+    RESOURCE_NOT_FOUND(null),
+    RETRY_ABORTED_BY_NO_AVAILABLE_REPLICA(AggRouterHttpRequestStats::recordNoAvailableReplicaAbortedRetryRequest);
+
+    final boolean reportUnhealthy;
+    final BiConsumer<AggRouterHttpRequestStats, String> statsRecorder;
+
+    FailureType(BiConsumer<AggRouterHttpRequestStats, String> statsRecorder) {
+      this.reportUnhealthy = statsRecorder == null;
```

Comment:
So the entire reportUnhealthy is used to recorded whether we have a non-null BiConsumer? Would it possible to just use 

```java
this.reportUnhealthy = Optional.ofNullable(statsRecorder)
    .orElse((stats, store) -> {});
```

Reply:
The `reportUnhealthy` is not only used within the constructor, it is also used inside the of `metricTracking` function to control some behavior.

Regarding the use of `Optional`, I personally do not find that it improves readability, so not a big fan... but that can be a matter of opinion 🤷

Reply:
I see.